src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientMethodsAnalyzer.cs

```diff
@@ -61,20 +72,158 @@ private static void CheckIsLastArgumentCancellationTokenOrRequestContext(ISymbol
+        private static void CheckProtocolMethodReturnType(ISymbolAnalysisContext context, IMethodSymbol method)
+        {
+            bool IsValidPageable(ITypeSymbol typeSymbol)
+            {
+                if ((!IsOrImplements(typeSymbol, PageableTypeName)) && (!IsOrImplements(typeSymbol, AsyncPageableTypeName)))
+                {
+                    return false;
+                }
+
+                var pageableTypeSymbol = typeSymbol as INamedTypeSymbol;
+                if (!pageableTypeSymbol.IsGenericType)
+                {
+                    return false;
+                }
+
+                var pageableReturn = pageableTypeSymbol.TypeArguments.Single();
+                if (!IsOrImplements(pageableReturn, BinaryDataTypeName))
+                {
+                    return false;
+                }
+
+                return true;
+            }
+
+            ITypeSymbol originalType = method.ReturnType;
+            ITypeSymbol unwrappedType = method.ReturnType;
+
+            if (method.ReturnType is INamedTypeSymbol namedTypeSymbol &&
+                namedTypeSymbol.IsGenericType &&
+                namedTypeSymbol.Name == TaskTypeName)
+            {
+                unwrappedType = namedTypeSymbol.TypeArguments.Single();
+            }
+
+            if (IsOrImplements(unwrappedType, ResponseTypeName))
+            {
+                if (unwrappedType is INamedTypeSymbol responseTypeSymbol && responseTypeSymbol.IsGenericType)
+                {
+                    var responseReturn = responseTypeSymbol.TypeArguments.Single();
+                    if (responseReturn.Name != BooleanTypeName)
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0018, method.Locations.FirstOrDefault()), method);
+                    }
+                }
+                return;
+            }
+            else if (IsOrImplements(unwrappedType, OperationTypeName))
+            {
+                if (unwrappedType is INamedTypeSymbol operationTypeSymbol && operationTypeSymbol.IsGenericType)
+                {
+                    var operationReturn = operationTypeSymbol.TypeArguments.Single();
+                    if (IsOrImplements(operationReturn, PageableTypeName) || IsOrImplements(operationReturn, AsyncPageableTypeName))
+                    {
+                        if (!IsValidPageable(operationReturn))
+                        {
+                            context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0018, method.Locations.FirstOrDefault()), method);
+                        }
+                        return;
+                    }
+
+                    if (!IsOrImplements(operationReturn, BinaryDataTypeName))
```

Comment:
Does this mean operations can only return `BinaryData` types?

Reply:
Yes. You could see all the accepted return types from the comments of this method:

> Response, Task<Response>, Response<bool>, Task<Response<bool>>, Pageable<BinaryData>, AsyncPageable<BinaryData>, Operation<BinaryData>, Task<Operation<BinaryData>>, Operation, Task<Operation>, Operation<Pageable<BinaryData>>, Task<Operation<AsyncPageable<BinaryData>>>